### PR TITLE
clues: dynamic viggora location

### PR DESCRIPTION
**Pull request by LlemonDuck**
Adds support for dynamic locations in the LocationClueScroll and LocationsClueScroll types, by exposing a ClueScrollPlugin instance as a parameter.

Also uses this to determine Viggora's location by varbit 815 and show the appropriate one.

Closes runelite#2142

### CODE REVIEW

**Comment by Felanbird** [comment]
<details>
<summary>Slayer Tower works as expected 👍</summary>

java_2023-07-09_17-27-52
java_2023-07-09_17-31-55

</details>

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClue.java [hunk 1]
```diff
@@ -334,12 +339,14 @@ public class CrypticClue extends ClueScroll implements NpcClueScroll, ObjectClue
 	private final String text;
 	private final String npc;
 	private final int objectId;
-	@Nullable
-	private final WorldPoint location;
 	private final String solution;
 	@Nullable
 	private final String questionText;
 
+	@Nullable
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClue.java:346]: This seems like it should be `Function<ClueScrollPlugin, @Nullable WorldPoint>`, not `@Nullable Function<ClueScrollPlugin, @Nullable WorldPoint>`. As-is, this introduces another field which must be null-checked

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClue.java [hunk 2]
```diff
@@ -500,4 +527,28 @@ public int[] getConfigKeys()
...
+		switch (varb)
+		{
+			case 1:
+				return VIGGORA_ROGUES_CASTLE;
+
+			case 2:
+				return VIGGORA_SLAYER_TOWER;
+
+			case 3:
+				return VIGGORA_EDGEVILLE_DUNGEON;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClue.java:547]: Note to self: need to verify this mapping

File: runelite-client/src/test/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClueTest.java [hunk 3]
```diff
@@ -24,14 +24,45 @@
...
 	}
+
+	@Test
+	public void forViggoraLocations()
+	{
+		when(plugin.getClient()).thenReturn(client);
+		when(client.getVarbitValue(Varbits.VIGGORA_LOCATION)).thenReturn(0, 1, 2, 3, 4);
+
+		CrypticClue clue = CrypticClue.forText("Come brave adventurer, your sense is on fire. If you talk to me, it's an old god you desire.");
+		assert clue != null;
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClueTest.java:60]: `assertNotNull(clue);`

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClue.java [hunk 4]
```diff
@@ -500,4 +527,28 @@ public int[] getConfigKeys()
 	{
 		return new int[]{text.hashCode()};
 	}
+
+	private static final WorldPoint VIGGORA_ROGUES_CASTLE = new WorldPoint(3295, 3934, 1);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClue.java:531]: usually, we put static fields at the top of classes

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClue.java (hunk 1 above)
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/clues/CrypticClue.java:346]: I think either way is ok. It is +1 extra null check in getLocation() below which seems fine.

**Pull request merged**